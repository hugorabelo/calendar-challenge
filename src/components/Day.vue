<template>
  <div
    class="flex flex-col box-border border border-gray-300 md:p-4 p-1 day-cell"
    :class="{ 'bg-blue-50': isWeekend, 'hover:border-blue-700': isAvailable }"
    @mouseover="upHere = true"
    @mouseleave="upHere = false"
  >
    <div class="relative w-full justify-items-center">
      <div class="relative h-8 text-lg text-black font-medium label-day" :class="{ 'text-gray-300': !isAvailable }">
        {{ dayLabel }}
      </div>
      <div class="w-full flex flex-col">
        <reminder-list :date="fullDate" :hovering="upHere"></reminder-list>
      </div>
    </div>
  </div>
</template>

<script>
import { DateTime } from 'luxon'
import { mapGetters } from 'vuex'
import ReminderList from './ReminderList.vue'

export default {
  name: 'Day',
  components: {
    ReminderList,
  },
  props: {
    fullDate: {
      type: Object,
      default: function () {
        return DateTime.now()
      },
    },
  },
  data() {
    return {
      upHere: false,
    }
  },
  computed: {
    ...mapGetters(['selectedMonth', 'reminders']),
    dayLabel() {
      return this.fullDate ? this.fullDate.toFormat('d') : ''
    },
    isAvailable() {
      return this.fullDate ? this.fullDate.toFormat('L') == this.selectedMonth.toFormat('L') : false
    },
    isWeekend() {
      return this.fullDate ? this.fullDate.toFormat('c') > 5 : false
    },
  },
}
</script>

<style>
.day-cell {
  height: calc((100vh - 100px) / 6) !important;
  overflow: hidden;
}
</style>
