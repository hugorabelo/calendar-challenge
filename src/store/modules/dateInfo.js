import { DateTime } from 'luxon'

const state = {
  selectedMonth: DateTime.now().startOf('month'),
}
const getters = {
  selectedMonth: (state) => state.selectedMonth,
}
const mutations = {
  setSelectedMonth(state, data) {
    state.selectedMonth = data
  },
}
const actions = {
  initSelectedMonth({ commit }) {
    commit('setSelectedMonth', DateTime.now().startOf('month'))
  },
  goToNextMonth({ commit, state, dispatch }) {
    if (!state.selectedMonth) {
      dispatch('initSelectedMonth')
    }
    let nextMonth = state.selectedMonth.plus({ months: 1 })
    commit('setSelectedMonth', nextMonth)
  },
  goToPreviousMonth({ commit, state, dispatch }) {
    if (!state.selectedMonth) {
      dispatch('initSelectedMonth')
    }
    let previousMonth = state.selectedMonth.minus({ months: 1 })
    commit('setSelectedMonth', previousMonth)
  },
}

export default {
  state,
  getters,
  actions,
  mutations,
}
