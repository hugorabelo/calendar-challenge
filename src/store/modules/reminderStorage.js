const state = {
  reminders: localStorage.getItem('reminders') != null ? JSON.parse(localStorage.getItem('reminders')) : [],
}
const getters = {
  reminders: (state) => state.reminders,
}
const mutations = {
  setReminders(state, data) {
    state.reminders = data
    localStorage.setItem('reminders', JSON.stringify(state.reminders))
  },
}
const actions = {
  addReminder({ commit, state }, data) {
    if (state.reminders.filter((rem) => rem.id == data.id).length == 0) {
      state.reminders.push(data)
    }
    commit('setReminders', state.reminders)
  },
  updateReminder({ commit, state }, data) {
    for (var i = 0; i < state.reminders.length; i++) {
      if (state.reminders[i].id == data.id) {
        state.reminders[i] = data
      }
    }
    commit('setReminders', state.reminders)
  },
  deleteReminder({ commit, state }, data) {
    for (var i = 0; i < state.reminders.length; i++) {
      if (state.reminders[i].id == data.id) {
        state.reminders.splice(i, 1)
      }
    }
    commit('setReminders', state.reminders)
  },
  deleteRemindersGroup({ dispatch }, data) {
    for (var i = 0; i < data.length; i++) {
      dispatch('deleteReminder', { id: data[i].id })
    }
  },
}

export default {
  state,
  getters,
  actions,
  mutations,
}
