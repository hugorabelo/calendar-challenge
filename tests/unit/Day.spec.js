import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Day from '@/components/Day'
import { DateTime } from 'luxon'
import dateInfo from '@/store/modules/dateInfo'
import reminderStorage from '@/store/modules/reminderStorage'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Day', () => {
  const build = (options) => {
    return mount(Day, {
      localVue,
      store: new Vuex.Store({
        modules: {
          dateInfo: { ...dateInfo },
          reminderStorage: { ...reminderStorage },
        },
      }),
      ...options,
    })
  }

  test('Is a Vue instance and valid date', () => {
    const { vm } = build()
    expect(vm).toBeTruthy()

    expect(DateTime.isDateTime(vm.fullDate)).toBeTruthy()
    expect(vm.fullDate.isValid).toBeTruthy()

    const dayOne = build({
      propsData: {
        fullDate: DateTime.fromISO('2022-13-03T00:00:00'),
      },
    })
    expect(DateTime.isDateTime(dayOne.vm.fullDate)).toBeTruthy()
    expect(dayOne.vm.fullDate.isValid).toBeFalsy()
  })

  test('Is day available', () => {
    const dayOne = build()
    expect(dayOne.vm.isAvailable).toBeTruthy()

    const dayTwo = build({
      propsData: {
        fullDate: DateTime.now().plus({ months: 1 }),
      },
    })
    expect(dayTwo.vm.isAvailable).toBeFalsy()
  })

  test('Is weekend', () => {
    const dayOne = build({
      propsData: {
        fullDate: DateTime.fromISO('2022-04-03T00:00:00'),
      },
    })
    expect(dayOne.vm.isWeekend).toBeTruthy()

    const dayTwo = build({
      propsData: {
        fullDate: DateTime.fromISO('2022-04-01T00:00:00'),
      },
    })
    expect(dayTwo.vm.isWeekend).toBeFalsy()
  })

  test('Get day label', () => {
    const dayOne = build()
    expect(dayOne.html()).toMatchSnapshot()

    const labelOne = dayOne.findComponent('.label-day')
    expect(labelOne.text()).toBe(dayOne.vm.fullDate.toFormat('d'))

    const dayTwo = build({
      propsData: {
        fullDate: DateTime.fromISO('2022-04-17T00:00:00'),
      },
    })
    const labelTwo = dayTwo.findComponent('.label-day')
    expect(labelTwo.text()).toBe('17')
  })
})
