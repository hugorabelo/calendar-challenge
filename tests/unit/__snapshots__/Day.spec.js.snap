// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day Get day label 1`] = `
<div class="flex flex-col box-border border border-gray-300 md:p-4 p-1 hover:border-blue-700">
  <div class="w-full justify-items-center">
    <div class="h-8 text-lg text-black font-medium label-day">
      30
    </div>
    <div class="w-full flex flex-col"></div>
  </div>
</div>
`;
